## Informe de Calidad de los datos "Metadata_sitios" proveniente de Google Maps:

In [1]:
#Librerias a utilizar
import pandas as pd
import ast
import geopy
from geopy.geocoders import Nominatim

 - Leer los archivo JSON en un DataFrame de Pandas (Primeras 6)

In [ ]:
metadata_01 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/1.json', lines=True)

In [ ]:
metadata_02 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/2.json', lines=True)

In [ ]:
metadata_03 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/3.json', lines=True)

In [ ]:
metadata_04 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/4.json', lines=True)

In [ ]:
metadata_05 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/5.json', lines=True)

In [ ]:
metadata_06 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/6.json', lines=True)

In [9]:
#Crear un diccionario que contenga los límites de latitud y longitud para el estado de Florida. 
locator = Nominatim(user_agent="myGeocoder")
location = locator.geocode("Florida, USA")

lat_min, lat_max = location.raw['boundingbox'][0], location.raw['boundingbox'][1]
lon_min, lon_max = location.raw['boundingbox'][2], location.raw['boundingbox'][3]

In [10]:
#pasar los datos a tipo float
lat_max = float(lat_max)
lat_min = float(lat_min)
lon_max = float(lon_max)
lon_min = float(lon_min)

- Filtrar solo las ubicaciones para florida

In [ ]:
metadata_01 = metadata_01[(metadata_01['latitude'] >= lat_min) & (metadata_01['latitude'] <= lat_max) & 
             (metadata_01['longitude'] >= lon_min) & (metadata_01['longitude'] <= lon_max)]

In [ ]:
metadata_02 = metadata_02[(metadata_02['latitude'] >= lat_min) & (metadata_02['latitude'] <= lat_max) & 
             (metadata_02['longitude'] >= lon_min) & (metadata_02['longitude'] <= lon_max)]

In [ ]:
metadata_03 = metadata_03[(metadata_03['latitude'] >= lat_min) & (metadata_03['latitude'] <= lat_max) & 
             (metadata_03['longitude'] >= lon_min) & (metadata_03['longitude'] <= lon_max)]

In [ ]:
metadata_04 = metadata_04[(metadata_04['latitude'] >= lat_min) & (metadata_04['latitude'] <= lat_max) & 
             (metadata_04['longitude'] >= lon_min) & (metadata_04['longitude'] <= lon_max)]

In [ ]:
metadata_05 = metadata_05[(metadata_05['latitude'] >= lat_min) & (metadata_05['latitude'] <= lat_max) & 
             (metadata_05['longitude'] >= lon_min) & (metadata_05['longitude'] <= lon_max)]

In [ ]:
metadata_06 = metadata_06[(metadata_06['latitude'] >= lat_min) & (metadata_06['latitude'] <= lat_max) & 
             (metadata_06['longitude'] >= lon_min) & (metadata_06['longitude'] <= lon_max)]

- Concatenar los dataframes usando la función concat (de 1 a 6)

In [ ]:
metadata_sitios_1 = pd.concat(
    [metadata_01, metadata_02, metadata_03, metadata_04, metadata_05, 
    metadata_06]
    )

In [ ]:
# Guardar el DataFrame en un archivo JSON
metadata_sitios_1.to_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/1_6.json', orient='records')

- Leer los archivo JSON en un DataFrame de Pandas (desde la 7 en adelante)

In [3]:
metadata_07 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/7.json', lines=True)

In [4]:
metadata_08 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/8.json', lines=True)

In [5]:
metadata_09 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/9.json', lines=True)

In [6]:
metadata_10 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/10.json', lines=True)

In [7]:
metadata_11 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/11.json', lines=True)

- Filtrar solo las ubicaciones para florida 

In [11]:
metadata_07 = metadata_07[(metadata_07['latitude'] >= lat_min) & (metadata_07['latitude'] <= lat_max) & 
             (metadata_07['longitude'] >= lon_min) & (metadata_07['longitude'] <= lon_max)]

In [12]:
metadata_08 = metadata_08[(metadata_08['latitude'] >= lat_min) & (metadata_08['latitude'] <= lat_max) & 
             (metadata_08['longitude'] >= lon_min) & (metadata_08['longitude'] <= lon_max)]

In [13]:
metadata_09 = metadata_09[(metadata_09['latitude'] >= lat_min) & (metadata_09['latitude'] <= lat_max) & 
             (metadata_09['longitude'] >= lon_min) & (metadata_09['longitude'] <= lon_max)]

In [14]:
metadata_10 = metadata_10[(metadata_10['latitude'] >= lat_min) & (metadata_10['latitude'] <= lat_max) & 
             (metadata_10['longitude'] >= lon_min) & (metadata_10['longitude'] <= lon_max)]

In [15]:
metadata_11 = metadata_11[(metadata_11['latitude'] >= lat_min) & (metadata_11['latitude'] <= lat_max) & 
             (metadata_11['longitude'] >= lon_min) & (metadata_11['longitude'] <= lon_max)]

- Concatenar los dataframes usando la función concat

In [16]:
metadata_sitios_2 = pd.concat(
    [metadata_07, metadata_08, metadata_09, metadata_10, metadata_11]
    )

In [17]:
# Guardar el DataFrame en un archivo JSON
metadata_sitios_2.to_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/7_11.json', orient='index')

- Leer los dos archivos JSON 

In [11]:
metadata_sitios_1 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/1_6.json')

In [12]:
metadata_sitios_2 = pd.read_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/7_11.json')

In [13]:
#Concatenar los dataframes usando la función concat
metadata_sitios = pd.concat(
    [metadata_sitios_1, metadata_sitios_2]
    )

In [15]:
# Guardar el DataFrame en un archivo JSON
metadata_sitios.to_json('/content/drive/MyDrive/Giselle PI/Proyecto Grupal/Google Maps/archivos metadata/metadata_sitios.json', orient='records')

In [18]:
# Vista
metadata_sitios.head()

name  \
0                        NTK OUTDOOR   
1      Cruises Inc. - Connie Stewart   
2         Seminole Family Restaurant   
3  First Impressions Barbershop Inc.   
4                  Brian Shaheen, MD   

                                             address  \
0  NTK OUTDOOR, 2315 NW 107th Ave #1B18, Miami, F...   
1  Cruises Inc. - Connie Stewart, 6602 52nd Ln, P...   
2  Seminole Family Restaurant, 6864 Seminole Blvd...   
3  First Impressions Barbershop Inc., 577 Deltona...   
4  Brian Shaheen, MD, 2421 Thomas Dr, Panama City...   

                                 gmap_id description   latitude  longitude  \
0   0x88d9beb4fe0532c1:0xef0555c169299d6        None  25.795204 -80.366038   
1  0x88c2e49b79f06c31:0x4ed8c8ce27e926e0        None  27.832187 -82.704805   
2  0x88c2fb8614e57ce9:0x40b51628875ad6e9        None  27.834885 -82.787419   
3  0x88e711bd2244fe3b:0x8406dd780f0574d1        None  28.877581 -81.281276   
4  0x8893863ea87bd5dd:0x9383ebf973e74abb        None  30.159982 -85.752277   

                                            category  avg_rating  \
0                                 [Corporate office]         5.0   
1                                    [Cruise agency]         5.0   
2                                               None         3.6   
3                                      [Barber shop]         4.0   
4  [Family practice physician, General practitioner]         4.2   

   num_of_reviews price                                              hours  \
0              35  None  [[Thursday, 9AM–6PM], [Friday, 9AM–6PM], [Satu...   
1               2  None  [[Thursday, 9AM–5PM], [Friday, 9AM–5PM], [Satu...   
2               8    $$                                               None   
3               8  None  [[Thursday, 8AM–8PM], [Friday, 8AM–8PM], [Satu...   
4              18  None  [[Thursday, 8AM–5PM], [Friday, 8AM–5PM], [Satu...   

                                                MISC              state  \
0  {'Accessibility': ['Wheelchair accessible entr...  Open ⋅ Closes 6PM   
1                                               None  Open ⋅ Closes 5PM   
2                                               None               None   
3  {'Accessibility': ['Wheelchair accessible entr...  Open ⋅ Closes 8PM   
4  {'Service options': ['Online care'], 'Accessib...  Open ⋅ Closes 5PM   

                                    relative_results  \
0  [0x88d9bfd49a624c3f:0x280afc17e8164d79, 0x88d9...   
1  [0x88c2fd3f88a7ab09:0x56dad3c1999b618f, 0x88c2...   
2  [0x88c2fc799a3fa1ad:0x2d1a46c8f525bf3e, 0x88c2...   
3  [0x88e7139dd918b543:0x2a27a69f19adfc9, 0x88e71...   
4  [0x88938ebfbd53f9c5:0xf6e52004f37523c8, 0x8893...   

                                                 url  
0  https://www.google.com/maps/place//data=!4m2!3...  
1  https://www.google.com/maps/place//data=!4m2!3...  
2  https://www.google.com/maps/place//data=!4m2!3...  
3  https://www.google.com/maps/place//data=!4m2!3...  
4  https://www.google.com/maps/place//data=!4m2!3...

In [14]:
# Columnas
metadata_sitios. columns

Index(['name', 'address', 'gmap_id', 'description', 'latitude', 'longitude',
       'category', 'avg_rating', 'num_of_reviews', 'price', 'hours', 'MISC',
       'state', 'relative_results', 'url'],
      dtype='object')

Lista que contiene las etiquetas de las columnas del DF "metadata_sitios", estos son:

name: Nombre del lugar o establecimiento.

address: Dirección del lugar o establecimiento.

gmap_id: ID del lugar o establecimiento en Google Maps.

description: Descripción del lugar o establecimiento.

latitude: Coordenada de latitud.

longitude: Coordenada de longitud.

category: Categoría y puede contener una lista de categorías separadas por coma.

avg_rating: Calificación promedio.

num_of_reviews: Número de reseñas que se han realizado.

price: Rango de precios si está disponible.

hours: Horario de atención en formato de lista.

MISC: Otros detalles relevantes como opciones de servicios adicionales que el lugar ofrece.

state: Estado en el que se encuentra el lugar.

relative_results: Resultados relacionados con la búsqueda.

url: URL del lugar o establecimiento en Google Maps.



In [16]:
# Total de datos foltrados por florida
metadata_sitios.shape

(229436, 15)

Contiene 229436 filas y 15 columnas.

In [17]:
# Obtener información sobre los tipos de datos, valores faltantes 
metadata_sitios.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 229436 entries, 0 to 105552
Data columns (total 15 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              229429 non-null  object 
 1   address           221067 non-null  object 
 2   gmap_id           229436 non-null  object 
 3   description       18359 non-null   object 
 4   latitude          229436 non-null  float64
 5   longitude         229436 non-null  float64
 6   category          227999 non-null  object 
 7   avg_rating        229436 non-null  float64
 8   num_of_reviews    229436 non-null  int64  
 9   price             18591 non-null   object 
 10  hours             176805 non-null  object 
 11  MISC              179589 non-null  object 
 12  state             179479 non-null  object 
 13  relative_results  204340 non-null  object 
 14  url               229436 non-null  object 
dtypes: float64(3), int64(1), object(11)
memory usage: 36.1+ MB


La descripción indica que se trata de un objeto DataFrame de Pandas con las siguientes características:

Int64Index: Tiene un índice de tipo Int64, lo que significa que cada fila está identificada por un número entero único. El índice va desde 0 hasta 105552 y tiene una longitud de 229436.

Data columns: Tiene un total de 15 columnas, cada una de las cuales contiene datos de diferentes tipos:

name: De tipo object.

address: De tipo object.

gmap_id: De tipo object.

description: De tipo object.

latitude: De tipo float64.

longitude: De tipo float64.

category: De tipo object.

avg_rating: De tipo float64.

num_of_reviews:De tipo int64.

price: De tipo object.

hours: De tipo object.

MISC:De tipo object.

state: De tipo object.

relative_results: De tipo object.

url: De tipo object.

Para cada columna, se muestra el número de valores no nulos, indicando que algunas columnas tienen valores faltantes (NaN).

memory usage: Indica el tamaño de memoria utilizado por el DataFrame que es de aproximadamente 36.1 megabytes.

In [19]:
# Estadísticas básicas de los datos
metadata_sitios.describe()

latitude      longitude     avg_rating  num_of_reviews
count  229436.000000  229436.000000  229436.000000   229436.000000
mean       27.870628     -81.723136       4.315791       52.423107
std         1.540787       1.553998       0.731479      180.201190
min        24.453973     -87.634653       1.000000        1.000000
25%        26.435938     -82.433105       4.000000        7.000000
50%        27.963251     -81.554113       4.500000       18.000000
75%        28.807684     -80.351286       4.900000       48.000000
max        31.000700     -80.006283       5.000000     9998.000000

Resumen estadístico de cuatro columnas del DataFrame "latitude, longitude, avg_rating y num_of_reviews". 

Algunas de las métricas y estadísticas que se pueden observar en este resumen son:

count: Número de valores no nulos para cada columna. Todas las columnas tienen el mismo número de valores no nulos.

mean: Promedio aritmético de los valores en cada columna. 

std: Desviación estándar de los valores en cada columna. 

min: Valor mínimo de cada columna.

25%, 50%, 75%: Percentiles correspondientes al 25%, 50% y 75% de los valores de cada columna.

max: Valor máximo de cada columna. 

In [20]:
# Datos nulos de cada columna 
metadata_sitios.isnull().sum()

name                     7
address               8369
gmap_id                  0
description         211077
latitude                 0
longitude                0
category              1437
avg_rating               0
num_of_reviews           0
price               210845
hours                52631
MISC                 49847
state                49957
relative_results     25096
url                      0
dtype: int64

Número de valores no nulos para cada columna en el DataFrame.

La columna "name" tiene 7 valores nulos; Esto representa una cantidad muy pequeña en comparación con el tamaño del DF, lo que sugiere que la información de nombres de la mayoría de los negocios se ha recopilado correctamente. Estos serán completados utilizando los datos de la columna "URL".

La columna "address" tiene 8369 valores nulos; Esto es más significativo en términos de tamaño del DF, lo que sugiere que hay muchos negocios para los cuales no se ha recopilado información de dirección. Se tendrá en cuenta eliminar estos nulos si es necesario.


La columna "description" tiene 211077 valores nulos; Esto significa que la mayoría de los negocios no tienen una descripción disponible en el conjunto de datos. Se considera eliminar la colummna.

La columna "category" tiene 1437 valores nulos; Aunque esto representa una cantidad relativamente pequeña de valores nulos, sigue siendo significativo dado el tamaño del DF. Se considera rellenar las columnas o eliminar los nulos.

La columna "price" tiene 210845 valores nulos; Lo que significa que la información sobre precios está disponible solo para una pequeña cantidad de negocios en el conjunto de datos. Se considera eliminar la columna.

La columna "hours" tiene 52631 valores nulos; Esto significa que la información de horas de operación está disponible solo para una fracción de los negocios. Se considera rellenar las columnas con "No detallado".

La columna "MISC" tiene 49847 valores nulos; Esto sugiere que la información de estado está disponible solo para una fracción de los negocios. Se decidirá luego de un analisis detallado.

La columna "state" tiene 49957 valores nulos; Esto sugiere que la información de estado está disponible solo para una fracción de los negocios. Se decidirá luego de un analisis detallado.

La columna "relative_results" tiene 25096 valores nulos; Esto sugiere que la información de estado está disponible solo para una fracción de los negocios. Se decidirá luego de un analisis detallado.

In [22]:
# Reemplazar las listas por cadena de caracteres
metadata_sitios['category'] = metadata_sitios['category'].apply(lambda x: str(x).strip('[]'))
metadata_sitios['hours'] = metadata_sitios['hours'].apply(lambda x: str(x).strip('[]'))
metadata_sitios['MISC'] = metadata_sitios['MISC'].apply(lambda x: str(x).strip('[]'))
metadata_sitios['relative_results'] = metadata_sitios['relative_results'].apply(lambda x: str(x).strip('[]'))

In [23]:
# Encunetro duplicados
metadata_sitios.duplicated().sum()

1854

Se encuentran 1854 duplicados, que despues de ser analizados en profundidad se decidirá si elminarlos o fucionarlos.